controllers/clustergroupupgrade_controller.go

```diff
@@ -973,14 +973,93 @@ func (r *ClusterGroupUpgradeReconciler) reconcileResources(ctx context.Context,
 	return nil
 }
 
-func (r *ClusterGroupUpgradeReconciler) buildRemediationPlan(ctx context.Context, clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade) {
+/* A policy's status has the format below. In this function we are creating a list with the clusternames
+   which are NonCompliant.
+
+status: (map[string]interface{})
+  compliant: NonCompliant
+  placement:
+  - placementBinding: operator-placementbinding
+    placementRule: operator-placementrules
+  status: ([]interface{})
+  - clustername: spoke3
+    clusternamespace: spoke3
+    compliant: NonCompliant
+  - clustername: spoke4
+    clusternamespace: spoke4
+    compliant: NonCompliant
+*/
+func (r *ClusterGroupUpgradeReconciler) getClustersNonCompliantWithAllManagedPolicies(
+	ctx context.Context, clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade, managedPolicies []*unstructured.Unstructured) map[string]bool {
+	clusterMap := make(map[string]bool)
+
+	// clustersNonCompliant will be a map of the clusters present in the CR and within how many managedPolicies they
+	// appear as being NonCompliant.
+	clustersNonCompliant := make(map[string]int)
+	for _, clusterName := range clusterGroupUpgrade.Spec.Clusters {
+		clustersNonCompliant[clusterName] = 0
+	}
+
+	for _, managedPolicy := range managedPolicies {
+		statusObject := managedPolicy.Object["status"].(map[string]interface{})
+		statusCompliance := statusObject["status"]
+
+		if statusCompliance == nil {
+			continue
+		}
+
+		for key, value := range statusObject {
+			r.Log.Info("key and value", "key", key, "value", value)
+		}
+
+		subStatus := statusCompliance.([]interface{})
+
+		for subStatusIndex := range subStatus {
+			crtSubStatus := subStatus[subStatusIndex].(map[string]interface{})
+			// If the cluster is NonCompliant in this current policy and it's also present in the list of clusters from the CR
+			// increment the number of policies for which it's NonCompliant.
+			if crtSubStatus["compliant"] == "NonCompliant" {
+				if _, ok := clustersNonCompliant[crtSubStatus["clustername"].(string)]; ok {
+					clustersNonCompliant[crtSubStatus["clustername"].(string)]++
+				}
+			}
+		}
+	}
+
+	for clusterName, clusterNonCompliantNumber := range clustersNonCompliant {
+		if clusterNonCompliantNumber == len(managedPolicies) {
+			clusterMap[clusterName] = true
+		} else {
+			clusterMap[clusterName] = false
+		}
+	}
+	r.Log.Info("Cluster (non) compliant with all managedPolicies map  : ", "clustersNotCompliantMap", clusterMap)
+	return clusterMap
+}
+
+func (r *ClusterGroupUpgradeReconciler) buildRemediationPlan(ctx context.Context, clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade, managedPolicies []*unstructured.Unstructured) {
+	clusterNonCompliantWithManagedPoliciesMap := r.getClustersNonCompliantWithAllManagedPolicies(ctx, clusterGroupUpgrade, managedPolicies)
```

Comment:
Does this mean that we only remediate clusters where ALL the managedPolicies are non compliant? My understanding is that the operator needs to remediate any cluster where any managedPolicy is non compliant.

Reply:
The operator needs to reconcile any non-compliant policy.

Reply:
Thank you for pointing this out, @imiller0. I'll update this.

Reply:
I believe so as well, the operator need to reconcile any non-compliant policy with the given clusters. Is there a reason that will make issue ?